Firmware/Marlin_main.cpp

```diff
@@ -3958,17 +3958,17 @@ static uint8_t get_PRUSA_SN(char* SN)
         {
             if (timeout.expired(250u))
                 goto exit;
-            if (MYSERIAL.available() > 0)
-            {
-                SN[rxIndex] = MYSERIAL.read();
-                rxIndex++;
-            }
+            int c = MYSERIAL.read();
+            if (c >= 0)
+                SN[rxIndex++] = (char)c;
         }
         SN[rxIndex] = 0;
         // printf_P(PSTR("SN:%s\n"), SN);
         SN_valid = (strncmp_P(SN, PSTR("CZPX"), 4) == 0);
     }
 exit:
+    _delay(100); // flush any remaining characters from the 32u2. Prevents enqueuing garbage data
+    MYSERIAL.flush(); //clear RX buffer
```

Comment:
I’m not sure if `flush()` clears the input buffer. It is my understanding that it will only block the program until active transmission is done and not guarantee an empty RX buffer.  

Perhaps using the MCU registers is more safe? Like this from Atmega2560 datasheet: 
![8C892175-6934-4012-BBCA-0A1F51122377](https://user-images.githubusercontent.com/8218499/149569405-cf9b6adf-7835-4a4c-b8bc-3e748ff0466b.jpeg)

Reply:
```
void MarlinSerial::flush()
{
  // don't reverse this or there may be problems if the RX interrupt
  // occurs after reading the value of rx_buffer_head but before writing
  // the value to rx_buffer_tail; the previous value of rx_buffer_head
  // may be written to rx_buffer_tail, making it appear as if the buffer
  // were full, not empty.
  rx_buffer.head = rx_buffer.tail;
}
```

In this firmware flush has a different meaning. It just means clear software RX buffer.
We could also delete the hardware buffer, but that should be cleared automatically by the RX interrupt.
Transmission is always blocking in the firmware (until the last byte is inserted in the TX hardware buffer, not really when that one is transmitted).

Reply:
I'm also a little scared by the flush function. The comment above it doesn't help since it contains conflicting information. As for the action itself, it should do the flush correctly, but I find it a bit risky. Maybe flush should be in a critical section so that the assignment is atomic